specutils/fitting/fitmodels.py

```diff
@@ -0,0 +1,507 @@
+        # representation of it which is a list that contains either astropy
+        # models or string representations of operators (e.g., '+' or '*').
+        model_in = [c.value for c in model_in._tree.traverse_postorder()]
+
+    #
+    # Run through each model in the list or compound model
+    #
+
+    model_out_stack = []
+    for sub_model in model_in:
+
+        #
+        # If it is an operator put onto the stack and move on...
+        #
+
+        if not isinstance(sub_model, Model):
+            model_out_stack.append(sub_model)
+            continue
+
+        #
+        # Make a new instance of the class.
+        #
+
+        if isinstance(sub_model, models.PolynomialModel):
+            new_sub_model = sub_model.__class__(sub_model.degree)
+        else:
+            new_sub_model = sub_model.__class__()
+
+        #
+        # Now for each parameter in the model determine if a dispersion or
+        # flux type of unit, then convert to spectrum units and then get the value.
+        #
+
+        for pn in new_sub_model.param_names:
+
+            # This could be a Quantity or Parameter
+            a = getattr(sub_model, pn)
+
+            if hasattr(a, 'quantity') and a.quantity is not None:
+                q = getattr(sub_model, pn).quantity
+
+                #
+                # Convert the quantity to the spectrum's units, and then we will use
+                # the *value* of it in the new unitless-model.
+                #
+
+                if q.unit.is_equivalent(dispersion_unit, equivalencies=u.equivalencies.spectral()):
+                    quantity = q.to(dispersion_unit, equivalencies=u.equivalencies.spectral())
+
+                elif q.unit.is_equivalent(flux_unit, equivalencies=u.equivalencies.spectral_density(dispersion)):
+                    quantity = q.to(flux_unit, equivalencies=u.equivalencies.spectral_density(dispersion))
+
+                # The value must be a quantity in order to use setattr (below)
+                # as the setter requires a Quantity on the rhs if the parameter
+                # is already a value.
+                v = quantity.value
+            else:
+                v = getattr(sub_model, pn).value
+
+            #
+            # Add this information for the parameter name into the
+            # new sub model.
+            #
+
+            setattr(new_sub_model, pn, v)
+
+        # The new model now has unitless information in it but has
+        # been converted to spectral unit scale.
+        model_out_stack.append(new_sub_model)
+
+    # If a compound model we need to re-create it, otherwise
+    # it is a single model and we just get the first one (as
+    # there is only one).
+    if compound_model:
+        model_out = _combine_postfix(model_out_stack)
+    else:
+        model_out = model_out_stack[0]
+
+    return model_out, dispersion.value, flux.value
+
+
+def _add_units_to_model(model_in, model_orig, spectrum):
```

Comment:
From here on we're getting into real code-sorcery level - nice work!

That said, I don't think I was able to give it a proper review, but I'll trust it since it seems to work... 😨 - hopefully we can clean it up a lot whenever we can get this fixed in Astropy. though.

Reply:
Sorcery is right.  Phew.

It works for what I give it, maybe another way to review for now is if you can think of other *tests* that should be added to make sure my code works. That way we have a better test base for all this. If you have ideas, feel free to add the ideas here or email/slack and I'll add them in.